src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -63,68 +76,120 @@ void BatchNorm<InputDataType, OutputDataType>::Reset()
 template<typename InputDataType, typename OutputDataType>
 template<typename eT>
 void BatchNorm<InputDataType, OutputDataType>::Forward(
-    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+    const arma::Mat<eT>& input,
+    arma::Mat<eT>& output)
 {
-  // Mean and variance over the entire training set will be used to compute
-  // the forward pass when deterministic is set to true.
-  if (deterministic)
+  Log::Assert(input.n_rows % size == 0, "Input features must be divisible \
+      by feature maps.");
+
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = input.n_rows / size;
+
+  // Set size of output equal to the size of input.
+  output.set_size(arma::size(input));
+
+  // We will calculate minibatch norm on each channel / feature map.
+  if (!deterministic)
   {
-    // Normalize the input and scale and shift the output.
-    output = input.each_col() - runningMean;
-    output.each_col() %= gamma / arma::sqrt(runningVariance / count + eps);
-    output.each_col() += beta;
+    // Input corresponds to output from convolution layer.
+    // Use a cube for simplicity.
+    arma::cube inputTemp(const_cast<arma::Mat<eT>&>(input).memptr(),
+        inputSize, size, batchSize, false, false);
+
+    // Initialize output to same size and values for convenience.
+    arma::cube outputTemp(const_cast<arma::Mat<eT>&>(output).memptr(),
+        inputSize, size, batchSize, false, false);
+    outputTemp = inputTemp;
+
+    // Calculate mean and variance over all channels.
+    mean = arma::mean(arma::mean(inputTemp, 2), 0);
+    variance = arma::mean(arma::mean(arma::pow(
+        inputTemp.each_slice() - arma::repmat(mean,
+        inputSize, 1), 2), 2), 0);
+
+    outputTemp.each_slice() -= arma::repmat(mean, inputSize, 1);
+
+    // Used in backward propagation.
+    inputMean.set_size(arma::size(inputTemp));
+    inputMean = outputTemp;
+
+    // Normalize output.
+    outputTemp.each_slice() /= arma::sqrt(arma::repmat(variance,
+        inputSize, 1) + eps);
+
+    // Re-used in backward propagation.
+    normalized.set_size(arma::size(inputTemp));
+    normalized = outputTemp;
+
+    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+        inputSize, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        inputSize, 1);
+
+    count += 1;
+    averageFactor = average ? 1.0 / count : momentum;
+
+    // Update running mean and running variance.
+    runningMean = (1 - averageFactor) * runningMean + averageFactor *
+        mean.t();
+    runningVariance = (1 - averageFactor) * runningVariance +
+       input.n_elem * (1.0 / (input.n_elem - size + eps)) *
+       averageFactor * variance.t();
```

Comment:
can you also explain where you are getting the `input.n_elem - size` from?
can you provide a link to the implementation?

Reply:
> and this doesn't (causes input stream error).
runningVariance = (1 - averageFactor) * runningVariance +
       input.n_elem * (1.0 / (input.n_elem - size + eps)) *
       averageFactor * variance.t();

This is the same as above? So, why do we not have the error?

Reply:
That's simplification of n / n - 1 where n = input.n_elem / size.

Reply:
We get an error if I remove eps.

Reply:
The reference implementation is same as above (The notebook and PyTorch).

Reply:
Also, I ran the tests 1500 times (I think twice or thrice) and it didn't fail so I think this works.

Reply:
for the linear layer this translates to `batchSize/(batchSize - 1)`. Are you sure this is correct? 
The implementation is for `batchnorm2d`. Does that apply to the linear layer?

Reply:
I think they are the same, Maybe this make [sense](https://github.com/pytorch/pytorch/issues/2628).

Reply:
This also might be [useful - PyTorch Discuss Forum](https://discuss.pytorch.org/t/why-2d-batch-normalisation-is-used-in-features-and-1d-in-classifiers/88360/2).

Reply:
No, my doubt is that if `batchSize = 1` then `batch/(batchSize- 1 + eps)` would be a very large number.

Reply:
Hmm, BatchSize number of columns. For one that would mean a single input point is passed through the dataset. How would you suggest to change. It does make sense that if batchSize == 1 then it would be a large number.

Reply:
So for that it would mean a shape would be 1x1.

Reply:
why does that mean that the number of rows would be 1? For example if the output size of a linear layer is 10 and the batchSize is 1 and we initialise batchnorm with size 10, then we would have output of 10x1 passed to batchNorm and n would be be 10/10 = 1. Is that not correct?

Reply:
Right, sorry that makes sense. I just tested would happen in pytorch with batchnorm 2d and the running var became nan,
```
A = np.array([ 5.1 , 3.5 , 1.4, 4.9 , 3.0 , 1.4 , 4.7 , 3.2 , 1.3 ]).reshape((1, 9, 1, 1))
my_bn = MyBatchNorm2d(9, eps = 1e-5, affine=True)
train_and_eval(my_bn, A)
```

Output : 
```
TRAIN_BN_RUNNING_VAR:
tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan])
```

Reply:
I think I got your point, I will look into the batchNorm 1d implementation as well and get back to you.

Reply:
Hey @saksham189, Could you also please look at [this thread](https://github.com/pytorch/pytorch/issues/7716).

Reply:
If this makes sense, I can push a commit to check the same. Kindly let me know what you think.

Reply:
Ahh, The serialisation test also had batchSize == 1 that's why there was an error.

Reply:
I have added a warning for now and set it 0 for batch size == 1 case. Kindly let me know what you think.

Reply:
yup looks fine now. Thanks for looking into this.

Reply:
Thanks for pointing it out, I'll push a commit for the style fix in just a sec.